### PR TITLE
Updated audiobook player to support new Audible TOC

### PR DESCRIPTION
**Pull request by nunommts**
**What's this do?**
This PR updates the logic to handle manifests with the `toc` object in order to update the app to support the new Audible TOC format. This new TOC's chapters may be related to more than one chapter, so the duration of the chapters and tracks may not be the same because one chapter may end and the track can still be played and a track may end and the chapter is not completed yet. After that, there's also a refactor on the `LCPAudioBookPlayer.kt` file in order to support the implemented changes and a change on the `PlayerPosition.kt` class to support the new field.

**Why are we doing this? (w/ JIRA link if applicable)**
There's a [ticket](https://www.notion.so/lyrasis/Update-Android-audiobook-player-to-support-Audible-TOC-0307c14b2f404b708d5744e6473d3187) saying the Audible has a new TOC format and the apps should be updated to match the new requirements.

**How should this be tested? / Do these changes have associated tests?**
Open the Palace app
Select "LYRASIS Reads" library
Search for the audiobook "The Martian" by Andy Weir
Play it and confirm everything's ok with the commands (seekbar dragging, play/pause, go back/forward, etc.)
Confirm if a new track is being played in the last 2 seconds of the "" chapter and nothing breaks

**Dependencies for merging? Releasing to production?**
[This PR](https://github.com/ThePalaceProject/android-core/pull/174) needs to be merged to support these changes

**Have you updated the changelog?**
Yes

**Has the application documentation been updated for these changes?**
No

**Did someone actually run this code to verify it works?**
Tested by @nunommts

### CODE REVIEW

**Review by ray-lee — CHANGES_REQUESTED** [review]
@nunommts There is a failing test. Also, could you add a test that uses the Anna Karenina manifest, and shows what is expected? That will help me to understand what this does.

**Comment by nunommts** [comment]
@ray-lee That's exactly what I'm trying to do right now, but I'm facing some issues while trying to run the tests. I'll update the PR when I fix them.

File: org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt [hunk 1]
```diff
@@ -242,637 +110,800 @@ class LCPAudioBookPlayer private constructor(
...
+
+    // the offset is greater than the track duration, so we need to get the next track
+    if (trackOffset > trackDurationMillis) {
+      setNextTrackToPlay(
+        index = currentTrackIndex + 1,
+        currentOffset = trackOffset - trackDurationMillis
+      )
+    } else if (trackOffset < 0) {
+
+      // the offset is fewer than 0, so we need to get the next track
```
Comment (ray-lee) [inline org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt:440]: I think you mean "get the previous track".

File: org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt [hunk 2]
```diff
@@ -242,637 +110,800 @@ class LCPAudioBookPlayer private constructor(
...
+  private fun cancelCurrentObserver() {
+    playbackObserver?.cancel(true)
+    spineElementToUpdate = null
   }
 
-  private fun handleActionOnSpineElement(
-    element: LCPSpineElement,
-    offset: Long,
-    playAutomatically: Boolean
+  private fun checkIfTrackHasChanged(
```
Comment (ray-lee) [inline org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt:410]: I think this function name is somewhat misleading. To me a "check" function returns a boolean depending on a condition, or possibly throws an exception if a condition isn't met. The main purpose of this function seems to be to update `trackIndex` (and optionally to seek to a position, and optionally start playing). How about naming it `updateTrackIndex`?

**Review by ray-lee — COMMENTED** [review]
This looks very good. I just noticed a few small things. 

Also (not relevant to this PR), does the player in android-drm-audioengine need to be updated?

File: org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt (hunk 1 above)
Comment (nunommts) [reply org.librarysimplified.audiobook.lcp/src/main/java/org/librarysimplified/audiobook/lcp/LCPAudioBookPlayer.kt:440]: You're right! Thanks for noticing

File: org.librarysimplified.audiobook.api/src/main/java/org/librarysimplified/audiobook/api/PlayerPosition.kt [hunk 3]
```diff
@@ -8,5 +8,6 @@ data class PlayerPosition(
   val title: String?,
   val part: Int,
   val chapter: Int,
-  val offsetMilliseconds: Long
+  val startOffset: Long,
+  val currentOffset: Long
```
[the remark this answers is not in the document]
Comment (nunommts) [reply org.librarysimplified.audiobook.api/src/main/java/org/librarysimplified/audiobook/api/PlayerPosition.kt:12]: Done. Please let me know if it's good.

**Comment by nunommts** [comment]
@ray-lee Good call! Yes, it does. I'll update it and create a PR.

**Review by ray-lee — APPROVED** [review]
Looks good!

**Comment by ray-lee** [comment]
@nunommts One more thing: This changes the API of PlayerPosition in a way that is not backward compatible, so it should be a major version upgrade. Could you change the version to 10.0.0-SNAPSHOT? Update it in the changelog too.

**Review by ray-lee — APPROVED** [review]
Looks good!

**Pull request merged**